neo4j/internal/bolt/message_queue.go

```diff
@@ -0,0 +1,206 @@
+package bolt
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/db"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/log"
+	"net"
+)
+
+type messageQueue struct {
+	in               incoming
+	out              outgoing
+	callbacks        list.List // List[responseHandler]
+	targetConnection net.Conn
+	protocolVersion  boltVersion
+	err              error
+
+	onNextMessage    func()
+	onNextMessageErr func(error)
+}
+
+func newMessageQueue(
+	target net.Conn,
+	in incoming, out outgoing,
+	onNext func(),
+	onNextErr func(error)) messageQueue {
+
+	return messageQueue{
+		in:               in,
+		out:              out,
+		callbacks:        list.List{},
+		targetConnection: target,
+		onNextMessage:    onNext,
+		onNextMessageErr: onNextErr,
+	}
+}
+
+func (q *messageQueue) appendHello(protocolVersion boltVersion, hello, token map[string]any,
+	helloHandler responseHandler, logonHandler responseHandler) {
+	q.protocolVersion = protocolVersion
+	q.out.appendHello(hello)
+	q.enqueueCallback(helloHandler)
+	if q.protocolVersion.greaterThanOrEqual(5, 1) {
+		q.out.appendLogon(token)
+		q.enqueueCallback(logonHandler)
+	}
+}
+
+func (q *messageQueue) appendRoute(routingContext map[string]string, bookmarks []string, extras map[string]any, handler responseHandler) {
+	q.out.appendRoute(routingContext, bookmarks, extras)
+	q.enqueueCallback(handler)
+}
+
+func (q *messageQueue) appendBegin(meta map[string]any, handler responseHandler) {
+	q.out.appendBegin(meta)
+	q.enqueueCallback(handler)
+}
+
+func (q *messageQueue) appendRun(cypher string, params, meta map[string]any, fetchSize int,
```

Comment:
Similar to my comment on `appendHello`, I think `appendRun` should not be responsible for pipelining `PULL`, but bolt should probably do this explicitly.

Reply:
Fixed in 50fddd2123cf63c1a23fe58bec26223b1ae84f68